armory/art_experimental/attacks/snr_pgd.py

```diff
@@ -0,0 +1,147 @@
+from art.attacks.evasion import ProjectedGradientDescentPyTorch
+import numpy as np
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class SNR_PGD(ProjectedGradientDescentPyTorch):
+    """
+    Applies L2 PGD to signal based on an SNR bound defined by norm 'snr' or 'snr_db'.
+        This is a *lower* bound on allowable SNR (as opposed to L2 upper bound)
+
+    norm - snr or snr_db
+    eps - the lower bound on allowable SNR and SNR_DB
+    eps_step - float value in (0, 2] that is the ratio of max L2 distance per step
+        A value of 2 is the diameter of the epsilon ball.
+        NOTE: this is different from original PGD because SNR is not additive
+
+    SNR measures the original signal power to adversarial perturbation power
+        If there is no perturbation, SNR is np.inf; all other values are finite.
+        If the perturbation is unbounded, SNR is 0
+    SNR_DB = 10 * log10(SNR)
+
+    If SNR is set to 0 or SNR_DB to -inf, no limit is provided to PGD
+        If SNR or SNR_DB is set to inf, no perturbation is performed
+    """
+
+    def __init__(
+        self, estimator, norm="snr", eps=10, eps_step=0.5, batch_size=1, **kwargs
+    ):
+        if batch_size != 1:
+            raise NotImplementedError("only batch_size 1 supported")
+        super().__init__(estimator, norm=2, batch_size=1, **kwargs)
+
+        # Map to SNR domain
+        eps = float(eps)
+        if norm == "snr":
```

Comment:
I think the logic here might be reversed. If `norm == "snr"`, we're already on a linear scale so I think we want `snr = eps`, and in the case where `norm == "snr_db"`, we want to do the conversion `snr = 10 ** (eps / 10)` to go from dB to a linear scale.

Reply:
Done.